consensus/istanbul/core/commit.go

```diff
@@ -142,16 +152,19 @@ func (c *core) handleCheckedCommitForCurrentSequence(msg *istanbul.Message, comm
 	// If we already have a proposal, we may have chance to speed up the consensus process
 	// by committing the proposal without PREPARE messages.
 	// TODO(joshua): Remove state comparisons (or change the cmp function)
-	if numberOfCommits >= minQuorumSize && c.state.Cmp(StateCommitted) < 0 {
+	if numberOfCommits >= minQuorumSize && c.current.State().Cmp(StateCommitted) < 0 {
 		logger.Trace("Got a quorum of commits", "tag", "stateTransition", "commits", c.current.Commits)
 		c.commit()
-	} else if c.current.GetPrepareOrCommitSize() >= minQuorumSize && c.state.Cmp(StatePrepared) < 0 {
-		if err := c.current.CreateAndSetPreparedCertificate(minQuorumSize); err != nil {
+	} else if c.current.GetPrepareOrCommitSize() >= minQuorumSize && c.current.State().Cmp(StatePrepared) < 0 {
+		err := c.current.TransitionToPrepared(minQuorumSize)
+		if err != nil {
 			logger.Error("Failed to create and set preprared certificate", "err", err)
 			return err
 		}
+		// Process Backlog Messages
+		c.processBacklog()
```

Comment:
Why `processBacklog` separately rather than inside `TransitionTo*`? Wouldn't the latter be more consistent with the current behavior?

Reply:
`RoundState` doesn't have access to `core` for now. I agree it's not the best design, I'll try to figure it out on a later PR

Reply:
Agree with Mariano, I had the same reaction at first, but unless we expose all roundState methods as an interface on core, i prefer this